tests/h/services/bulk_executor/_executor_test.py

```diff
@@ -0,0 +1,108 @@
+from h.services.bulk_executor._executor import BulkExecutor
+from tests.h.services.bulk_executor.conftest import CommandFactory
+
+
+class TestDBExecutor:
+    @pytest.mark.parametrize(
+        "command_type,data_type,handler",
+        (
+            param(
+                CommandType.UPSERT,
+                DataType.USER,
+                "UserUpsertAction",
+                marks=pytest.mark.xfail(reason="Not implemented"),
+            ),
+            param(
+                CommandType.UPSERT,
+                DataType.GROUP,
+                "GroupUpsertAction",
+                marks=pytest.mark.xfail(reason="Not implemented"),
+            ),
+            param(
+                CommandType.CREATE,
+                DataType.GROUP_MEMBERSHIP,
+                "GroupMembershipCreateAction",
+                marks=pytest.mark.xfail(reason="Not implemented"),
+            ),
+        ),
+        indirect=["handler"],
+    )
+    def test_it_calls_correct_db_handler(
+        self, db_session, command_type, data_type, handler, commands
+    ):
+        executor = BulkExecutor(db_session)
+        handler.assert_called_once_with(db_session)
+
+        executor.effective_user_id = 1
+
+        # These commands aren't actually the right type, but it doesn't
+        # matter for this test
+        executor.execute_batch(
+            command_type, data_type, {"config_option": 2}, commands,
+        )
+
+        bulk_upsert = handler.return_value
+        bulk_upsert.execute.assert_called_once_with(
+            commands, effective_user_id=1, config_option=2
+        )
+
+    @pytest.mark.parametrize(
+        "command_type,data_type",
+        (
+            (CommandType.CREATE, DataType.USER),
+            (CommandType.CREATE, DataType.GROUP),
+            (CommandType.UPSERT, DataType.GROUP_MEMBERSHIP),
+        ),
+    )
+    def test_it_raises_UnsupportedOperationError_for_invalid_actions(
+        self, db_session, command_type, data_type, commands
+    ):
+        executor = BulkExecutor(db_session)
+
+        with pytest.raises(UnsupportedOperationError):
+            executor.execute_batch(command_type, data_type, commands, {})
+
+    def test_it_raises_InvalidDeclarationError_with_non_lms_authority(self):
+        config = Configuration.create(
+            effective_user="acct:user@bad_authority.com", total_instructions=2
+        )
+
+        with pytest.raises(InvalidDeclarationError):
+            BulkExecutor(sentinel.db).configure(config)
+
+    @pytest.mark.parametrize(
+        "command",
+        (
+            param(CommandFactory.user_upsert(authority="bad"), id="bad user attr"),
+            param(
+                CommandFactory.user_upsert(query_authority="bad"), id="bad user query"
+            ),
+            param(CommandFactory.group_upsert(authority="bad"), id="bad group attr"),
+            param(
+                CommandFactory.group_upsert(query_authority="bad"), id="bad group query"
```

Comment:
The params here make this ugly, but without you can't tell what failed as it just prints an obscure object reference

Reply:
I've updated these with more descriptive values.